nemo/collections/asr/parts/k2/utils.py

```diff
@@ -0,0 +1,205 @@
+    result = k2.intersect(k2.arc_sort(base_graph), aux_graph_with_self_loops, treat_epsilons_specially=False,)
+    setattr(result, "phones", result.labels)
+    return result
+
+
+def compose_with_self_loops(base_graph: k2.Fsa, aux_graph: k2.Fsa) -> k2.Fsa:
+    """Composition helper function.
+    """
+    aux_graph_with_self_loops = k2.arc_sort(k2.add_epsilon_self_loops(aux_graph)).to(base_graph.device)
+    return k2.compose(base_graph, aux_graph_with_self_loops, treat_epsilons_specially=False, inner_labels="phones",)
+
+
+def create_sparse_wrapped(
+    indices: List[torch.Tensor],
+    values: torch.Tensor,
+    size: Optional[Union[Tuple[int, int], Tuple[int, int, int]]] = None,
+    min_col_index: Optional[int] = None,
+) -> torch.Tensor:
+    """Wraps up k2.create_sparse to create 2- or 3-dimensional sparse tensors.
+    """
+    assert size is None or len(indices) == len(size)
+
+    if len(indices) == 2:
+        return k2.create_sparse(
+            rows=indices[0], cols=indices[1], values=values, size=size, min_col_index=min_col_index,
+        )
+    elif len(indices) == 3:
+        assert indices[0].ndim == indices[1].ndim == indices[2].ndim == 1
+        assert indices[0].numel() == indices[1].numel() == indices[2].numel() == values.numel()
+
+        if min_col_index is not None:
+            assert isinstance(min_col_index, int)
+            kept_indices = indices[-1] >= min_col_index
+            indices = [i[kept_indices] for i in indices]
+            values = values[kept_indices]
+        if size is not None:
+            return torch.sparse_coo_tensor(
+                torch.stack(indices), values, size=size, device=values.device, requires_grad=values.requires_grad,
+            )
+        else:
+            return torch.sparse_coo_tensor(
+                torch.stack(indices), values, device=values.device, requires_grad=values.requires_grad,
+            )
+    else:
+        raise ValueError(f"len(indices) = {len(indices)}")
+
+
+def prep_padded_densefsavec(log_softmax: torch.Tensor, supervisions: torch.Tensor) -> k2.DenseFsaVec:
+    """Performs special epsilon-padding required for composition with some of the topologies.
+    """
+    log_softmax_shifted = torch.cat(
+        [
+            torch.full((log_softmax.shape[0], log_softmax.shape[1], 1), -float("inf"), device=log_softmax.device,),
+            log_softmax,
+        ],
+        axis=-1,
+    )
+    log_softmax_padded = torch.zeros(
+        (log_softmax_shifted.shape[0], log_softmax_shifted.shape[1] * 2, log_softmax_shifted.shape[2],),
+        device=log_softmax.device,
+    )
+    log_softmax_padded[:, ::2] = log_softmax_shifted
+    supervisions_padded = supervisions.clone()
+    supervisions_padded[:, 2] *= 2
+    dense_log_softmax_padded = k2.DenseFsaVec(log_softmax_padded, supervisions_padded)
+    return dense_log_softmax_padded
+
+
+def shift_labels_inpl(lattices: List[k2.Fsa], shift: int):
```

Comment:
Same as above

Reply:
Done in [7b2b4a4](https://github.com/NVIDIA/NeMo/pull/3351/commits/7b2b4a4ba4e34848d526d8c4ac53903bfbfe2fd2)